arch/cpu/nrf/sys/lpm.c

```diff
@@ -0,0 +1,293 @@
+ */
+static void
+wake_up(void)
+{
+
+  ENERGEST_SWITCH(ENERGEST_TYPE_DEEP_LPM, ENERGEST_TYPE_CPU);
+
+  watchdog_periodic();
+
+  /*
+   * In LowPower Mode, HFCLK is used by the IEEE 802.15.4 radio.
+   * and NRF_RTIMER_TIMER is used in the nrf-ieee-driver-arch.c
+   * In non-LowPower Mode, HFCLK is also used for the RTIMER
+   */
+  nrfx_clock_hfclk_start();
+  nrf_timer_task_trigger(NRF_RTIMER_TIMER, NRF_TIMER_TASK_START);
+}
+/*---------------------------------------------------------------------------*/
+static uint8_t
+check_next_rtimer(rtimer_clock_t now, rtimer_clock_t *next_rtimer, bool *next_rtimer_set)
+{
+  uint8_t max_pm = LPM_MODE_MAX_SUPPORTED;
+
+  if(rtimer_arch_next()) {
+    *next_rtimer_set = true;
+
+    /* find out the timer of the next rtimer interrupt */
+    *next_rtimer = rtimer_arch_next();
+
+    if(RTIMER_CLOCK_LT(*next_rtimer, now + 2)) {
+      max_pm = MIN(max_pm, LPM_MODE_AWAKE);
+    } else if(RTIMER_CLOCK_LT(*next_rtimer, now + STANDBY_MIN_DURATION)) {
+      max_pm = MIN(max_pm, LPM_MODE_SLEEP);
+    }
+  } else {
+    *next_rtimer_set = false;
+  }
+
+  return max_pm;
+}
+/*---------------------------------------------------------------------------*/
+static uint8_t
+check_next_etimer(rtimer_clock_t now, rtimer_clock_t *next_etimer, bool *next_etimer_set)
+{
+  uint8_t max_pm = LPM_MODE_MAX_SUPPORTED;
+
+  *next_etimer_set = false;
+
+  /* Find out the time of the next etimer */
+  if (etimer_pending()) {
+    clock_time_t next = etimer_next_expiration_time();
+    *next_etimer_set = true;
+    *next_etimer = next * (RTIMER_SECOND / CLOCK_SECOND);
+    if (RTIMER_CLOCK_LT(*next_etimer, now + STANDBY_MIN_DURATION)) {
+      max_pm = MIN(max_pm, LPM_MODE_SLEEP);
+    }
+  }
+
+  return max_pm;
+}
+/*---------------------------------------------------------------------------*/
+static uint8_t
+setup_sleep_mode(void)
+{
+  uint8_t max_pm = LPM_MODE_MAX_SUPPORTED;
+  uint8_t pm;
+
+  rtimer_clock_t now;
+  rtimer_clock_t trigger_etimer = 0;
+  rtimer_clock_t next_rtimer = 0;
+  rtimer_clock_t next_etimer = 0;
+  bool next_rtimer_set = false;
+  bool next_etimer_set = false;
+
+  /* Check if any events fired before we turned interrupts off. If so, abort */
+  if(LPM_MODE_MAX_SUPPORTED == LPM_MODE_AWAKE || process_nevents()) {
+    return LPM_MODE_AWAKE;
+  }
+
+  now = RTIMER_NOW();
+  pm = check_next_rtimer(now, &next_rtimer, &next_rtimer_set);
+  if(pm < max_pm) {
+    max_pm = pm;
+  }
+  pm = check_next_etimer(now, &next_etimer, &next_etimer_set);
+  if(pm < max_pm) {
+    max_pm = pm;
+  }
+  trigger_etimer = RTIMER_CLOCK_MAX;
+  if(max_pm == LPM_MODE_SLEEP) {
+    /* In sleep mode, HFXO is powered up, therefore we do  not have
+     * to wakeup earlier to powering up the HFXO */
+    if(next_etimer_set) {
+      /* Schedule the next system wakeup due to etimer */
+      if(RTIMER_CLOCK_LT(next_etimer, now + MIN_SAFE_SCHEDULE)) {
+        /* Too soon in future, keep the system awake */
```

Comment:
We get warnings from empty if-blocks, would be nice if the logic is changed to avoid this branch.

Reply:
I normally do no empty if-blocks, but here it made sense to me, because it shows the algorithm better. Let me know, if you need to have this corrected